### ISSUE TITLE
Pirate ships re-use old orders

### ISSUE THREAD

**Issue by shofixtum** [endless-sky/endless-sky#3739]
Strange behavior for pirate ships, seen on steam version and a jenkins nightly build.

Steps to reproduce:

1. Dominate a pirate world with a shipyard.
2. Capture some small pirate ships in the dominated system.
3. Right click to give the new escort ships orders to move to a location in the system.
4. Land on the dominated planet.
5. Sell the escorts.
6. Depart from the planet.

Expected behavior:
New pirate ships launched from the planet should move to attack the player.

Observed behavior:
Some new pirate ships launched from the planet move to the previously ordered location from step 3 and stay there. They remain hostile to the player, and fire on you if you move in their range, but remain in the same spot.

**Comment by CyberShadow** [comment]
Hey, I think I've seen this with friendly ships too. When capturing ships in a system (capture, land, park all, repeat), some AI (merchant etc.) ships will congregate in coordinates used in previous orders.

**Comment by tehhowch** [comment]
A good find, and a likely case of heap memory reuse, e.g. [this](https://stackoverflow.com/questions/11161682/do-memory-addresses-on-the-heap-get-reused), [this](https://jira.mongodb.org/browse/SERVER-14198). 

For those who notice this behavior, can you share details of your environment? e.g. OS, CPU/mobo, game version/commit. For those who build themselves, can you verify if it occurs in both the release and debug configurations (`scons` and `scons mode=debug`)?

------
Some more information about this bug:

The ships in the game that are in flight are all held in a `std::list` by the Engine class: 
https://github.com/endless-sky/endless-sky/blob/ffd15a7453c7cc48b5c70ed7a8bbd3beb4f7e1cb/source/Engine.h#L137
The logical organization of the `ships` list is initially set by `Engine::Place` to the following
```
| < launch-spawned NPCs > | flagship ... last flying playership | oldest ... mission NPC ships .... newest | < in-flight spawns > |
```

The AI class uses this list directly:
https://github.com/endless-sky/endless-sky/blob/ffd15a7453c7cc48b5c70ed7a8bbd3beb4f7e1cb/source/AI.h#L48-L50

Currently, escort orders are held in a `std::map` organized by a raw pointer to the ship:
https://github.com/endless-sky/endless-sky/blob/ffd15a7453c7cc48b5c70ed7a8bbd3beb4f7e1cb/source/AI.h#L189

The pointer used for storing/recalling orders is from the addressof operator `&`:
https://github.com/endless-sky/endless-sky/blob/ffd15a7453c7cc48b5c70ed7a8bbd3beb4f7e1cb/source/AI.cpp#L1040-L1042
and the `Ship& ship` passed to `FollowOrders` comes from dereferencing the stored pointer in the `shared_ptr` of `Engine`'s list: 
https://github.com/endless-sky/endless-sky/blob/ffd15a7453c7cc48b5c70ed7a8bbd3beb4f7e1cb/source/AI.cpp#L339
https://github.com/endless-sky/endless-sky/blob/ffd15a7453c7cc48b5c70ed7a8bbd3beb4f7e1cb/source/AI.cpp#L665

In every frame (in `Engine::CalculateStep`), ship nodes are erased in `Prune(ships)`, and then new ships are (occasionally) slated for addition in `SpawnFleets` (first added to an intermediate list), and then finally spliced back to the end of the ship list:
https://github.com/endless-sky/endless-sky/blob/ffd15a7453c7cc48b5c70ed7a8bbd3beb4f7e1cb/source/Engine.cpp#L1203-L1236

We don't see any strange behavior in random NPCs without landing & taking off because right now only player ships can be issued orders, and the likelihood of a player ship with an order being pruned in the same frame that a random NPC will be spawned is really low. (If we use `Orders` for #3381 , this could change). The reason we see it on takeoff after parking or selling is because there is a mass modification of the `ships` list in `Engine::Place`: the whole list is cleared, then random fleets are spawned, and then the player's & mission ships are added. Because the player's & mission ships are each managed elsewhere (`PlayerInfo->ships` and `NPC->ships`), clearing the `ships` list does not free the heap memory associated with these particular instances (a good thing). When they are added back to the `Engine->ships` list, they retain their original memory address (and thus giving orders persists through landing & takeoff, so long as the game is not reloaded). Random fleets, however, are created anew, and the OS correctly notes that it has some rather nicely sized memory spaces in which it can store `Ship` instance data (because that's what it was previously used for).


To remedy this situation, we could either change the `orders` map signature, be hyper-vigilant about erasing a ship's `Orders` if the ship gets zonked somehow, or totally change how orders are implemented.

**Comment by shofixtum** [comment]
I see the bug on Steam version 0.9.8, running on an Alienware M18X laptop, Intel Core i7 2720QM 2.2GHz, 6G RAM, with Windows 10 Home. It's easy for me to reproduce.

**Comment by endless-sky** [comment]
This should be (mostly) fixed by 9cdebdb. There's still a slight chance that you'll capture a ship, give it orders, then lose it, then capture a new ship with the same pointer, but it's unlikely.
